velox/expression/tests/ArgumentTypeFuzzerTest.cpp

```diff
@@ -403,4 +403,205 @@ TEST_F(ArgumentTypeFuzzerTest, orderableConstraint) {
+                  .integerVariable("s")
+                  .returnType("boolean")
+                  .argumentType("decimal(3,s)")
+                  .build();
+
+  argTypes = fuzzArgumentTypes(*signature, BOOLEAN());
+  ASSERT_EQ(1, argTypes.size());
+  EXPECT_TRUE(argTypes[0]->isDecimal());
+  EXPECT_EQ(3, getDecimalPrecisionScale(*argTypes[0]).first);
+
+  // Another way to specify fixed precision.
+  signature = exec::FunctionSignatureBuilder()
+                  .integerVariable("p", "30")
+                  .integerVariable("s")
+                  .returnType("boolean")
+                  .argumentType("decimal(p,s)")
+                  .build();
+
+  argTypes = fuzzArgumentTypes(*signature, BOOLEAN());
+  ASSERT_EQ(1, argTypes.size());
+  EXPECT_TRUE(argTypes[0]->isDecimal());
+  EXPECT_EQ(30, getDecimalPrecisionScale(*argTypes[0]).first);
+
+  // Multiple arguments. All must be the same as return type.
+  signature = exec::FunctionSignatureBuilder()
+                  .integerVariable("p")
+                  .integerVariable("s")
+                  .returnType("decimal(p,s)")
+                  .argumentType("decimal(p,s)")
+                  .argumentType("decimal(p,s)")
+                  .argumentType("decimal(p,s)")
+                  .build();
+
+  argTypes = fuzzArgumentTypes(*signature, DECIMAL(10, 7));
+  ASSERT_EQ(3, argTypes.size());
+  EXPECT_EQ(DECIMAL(10, 7)->toString(), argTypes[0]->toString());
+  EXPECT_EQ(DECIMAL(10, 7)->toString(), argTypes[1]->toString());
+  EXPECT_EQ(DECIMAL(10, 7)->toString(), argTypes[2]->toString());
+
+  // Multiple arguments. Some have fixed precision, scale or both.
+  signature = exec::FunctionSignatureBuilder()
+                  .integerVariable("p")
+                  .integerVariable("s")
+                  .returnType("decimal(p,s)")
+                  .argumentType("decimal(p,s)")
+                  .argumentType("decimal(p,10)")
+                  .argumentType("decimal(12,s)")
+                  .argumentType("decimal(2,1)")
+                  .build();
+
+  argTypes = fuzzArgumentTypes(*signature, DECIMAL(10, 7));
+  ASSERT_EQ(4, argTypes.size());
+  EXPECT_EQ(DECIMAL(10, 7)->toString(), argTypes[0]->toString());
+  EXPECT_EQ(DECIMAL(10, 10)->toString(), argTypes[1]->toString());
+  EXPECT_EQ(DECIMAL(12, 7)->toString(), argTypes[2]->toString());
+  EXPECT_EQ(DECIMAL(2, 1)->toString(), argTypes[3]->toString());
+}
+
+TEST_F(ArgumentTypeFuzzerTest, fuzzDecimalReturnType) {
+  auto fuzzReturnType = [](const exec::FunctionSignature& signature) {
+    std::mt19937 seed{0};
+    ArgumentTypeFuzzer fuzzer{signature, seed};
+    return fuzzer.fuzzReturnType();
+  };
+
+  // Return type can be any decimal.
+  auto signature = exec::FunctionSignatureBuilder()
+                       .integerVariable("p")
+                       .integerVariable("s")
+                       .returnType("decimal(p,s)")
+                       .argumentType("decimal(p,s)")
```

Comment:
Maybe we can add test case for varargs like greatest and least.

Reply:
Good idea. Added

```
  signature = exec::FunctionSignatureBuilder()
                  .integerVariable("p")
                  .integerVariable("s")
                  .returnType("decimal(p,s)")
                  .argumentType("decimal(p,s)")
                  .argumentType("decimal(p,s)")
                  .variableArity()
                  .build();
  argTypes = fuzzArgumentTypes(*signature, DECIMAL(10, 7));
  ASSERT_LE(1, argTypes.size());
  for (const auto& argType : argTypes) {
    EXPECT_EQ(DECIMAL(10, 7)->toString(), argType->toString());
  }
```